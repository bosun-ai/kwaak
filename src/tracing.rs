use crate::repository::Repository;
use anyhow::Result;
use tracing::level_filters::LevelFilter;
use tracing_opentelemetry::OpenTelemetryLayer;
use tracing_subscriber::prelude::*;
use tracing_subscriber::{fmt, EnvFilter};

// For now just log to stdout and file
pub fn init(repository: &Repository) -> Result<()> {
    let log_dir = repository.config().log_dir();

    let file_appender = tracing_appender::rolling::daily(
        log_dir,
        format!("{}.log", repository.config().project_name),
    );

    let fmt_layer = fmt::layer().with_writer(file_appender);

    // Logs the file layer will capture
    let env_filter_layer = EnvFilter::builder()
        .with_default_directive(LevelFilter::INFO.into())
        .from_env_lossy()
        .add_directive("h2=error".parse().unwrap())
        .add_directive("tower=error".parse().unwrap())
        .add_directive("tui_markdown=error".parse().unwrap());

    // The log level tui logger will capture
    let default_level = if cfg!(debug_assertions) {
        log::LevelFilter::Info
    } else {
        log::LevelFilter::Warn
    };
    let tui_layer = tui_logger::tracing_subscriber_layer();
    tui_logger::init_logger(default_level)?;

    let mut layers = vec![
        // env_filter_layer.boxed(),
        tui_layer.boxed(),
        fmt_layer.boxed(),
    ];

    if cfg!(feature = "otel") {
        dbg!("OpenTelemetry tracing enabled");
        let provider = otel_provider();
        let tracer = provider.tracer("kwaak");
        opentelemetry::global::set_tracer_provider(provider);

        // Create a tracing layer with the configured tracer
        let layer = OpenTelemetryLayer::new(tracer);

        layers.push(layer.boxed());
    }

    let registry = tracing_subscriber::registry()
        .with(env_filter_layer)
        .with(layers);
    registry.try_init()?;

    Ok(())
}

#[cfg(feature = "otel")]
use opentelemetry::trace::TracerProvider as _;
#[cfg(feature = "otel")]
use opentelemetry_sdk::trace::TracerProvider;

#[cfg(feature = "otel")]
fn otel_provider() -> TracerProvider {
    use opentelemetry_sdk::runtime;
    use opentelemetry_sdk::trace::TracerProvider;

    let exporter = opentelemetry_otlp::SpanExporter::builder()
        .with_tonic()
        .build()
        .expect("failed to create otlp exporter");

    TracerProvider::builder()
        .with_batch_exporter(exporter, runtime::Tokio)
        .with_config(opentelemetry_sdk::trace::Config::default().with_resource(
            opentelemetry_sdk::Resource::new(vec![opentelemetry::KeyValue::new(
                "service.name",
                "kwaak",
            )]),
        ))
        .build()
}
