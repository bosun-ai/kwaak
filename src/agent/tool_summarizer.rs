use std::sync::Arc;

use swiftide::{
    agents::hooks::AfterToolFn,
    chat_completion::{Tool, ToolCall, ToolOutput},
    prompt::Prompt,
    traits::{Command, SimplePrompt},
};
use tracing::Instrument as _;

use crate::{templates::Templates, util::accept_non_zero_exit};

#[derive(Clone)]
pub struct ToolSummarizer<'a> {
    llm: Arc<dyn SimplePrompt>,
    tools_to_summarize: Vec<&'a str>,
    available_tools: Vec<Box<dyn Tool>>,
    git_start_sha: String,
}

impl<'a> ToolSummarizer<'a> {
    pub fn new(
        llm: Box<dyn SimplePrompt>,
        tools_to_summarize: &[&'a str],
        available_tools: &[Box<dyn Tool>],
        git_start_sha: impl Into<String>,
    ) -> Self {
        Self {
            llm: llm.into(),
            tools_to_summarize: tools_to_summarize.into(),
            available_tools: available_tools.into(),
            git_start_sha: git_start_sha.into(),
        }
    }

    // Rust has a long outstanding issue where it captures outer lifetimes when returning an impl
    // that also has lifetimes, see https://github.com/rust-lang/rust/issues/42940
    pub fn summarize_hook(self) -> impl AfterToolFn + use<'a> {
        move |agent, tool_call, tool_output| {
            let llm = self.llm.clone();

            let Some(tool) = self
                .tools_to_summarize
                .iter()
                .find(|t| *t == &tool_call.name())
                .and_then(|t| self.available_tools.iter().find(|tool| &tool.name() == t))
            else {
                return Box::pin(async move { Ok(()) });
            };

            if let Ok(output) = tool_output {
                let git_start_sha = self.git_start_sha.clone();

                let span = tracing::info_span!("summarize_tool", tool = tool.name().as_ref());
                let prompt = prompt(tool, tool_call, output, self.available_tools.as_slice());

                return Box::pin(
                    async move {
                        let current_diff = accept_non_zero_exit(
                            agent.context()
                                .exec_cmd(&Command::shell(format!(
                                    "git diff {git_start_sha} --no-color"
                                )))
                                .await,
                        )?
                        .output;

                        let current_diff = if current_diff.is_empty() {
                            None
                        } else {
                            Some(current_diff)
                        };

                        let prompt = prompt.with_context_value("diff", current_diff);

                        let summary = llm.prompt(prompt).await?;
                        tracing::debug!(summary = %summary, original = %output, "Summarized tool output");
                        *output = summary.into();

                        Ok(())
                    }
                    .instrument(span),
                );
            }
            Box::pin(async move { Ok(()) })
        }
    }
}

fn prompt(
    tool: &dyn Tool,
    tool_call: &ToolCall,
    tool_output: &ToolOutput,
    available_tools: &[Box<dyn Tool>],
) -> Prompt {
    // TODO: Partial?
    let formatted_tools = available_tools
        .iter()
        .map(|tool| {
            format!(
                "- **{}**: {}",
                tool.name(),
                // Some tools have large descriptions, only take the first line
                tool.tool_spec()
                    .description
                    .lines()
                    .take(1)
                    .collect::<String>()
            )
        })
        .collect::<Vec<String>>()
        .join("\n");

    // TODO: clean me up further
    let template =
        Templates::load("tool_summarizer.md").expect("Infallible; failed to find template");

    template
        .to_prompt()
        .with_context_value("formatted_tools", formatted_tools)
        .with_context_value("tool_name", tool.name())
        .with_context_value("tool_description", tool.tool_spec().description)
        .with_context_value("tool_args", tool_call.args())
        .with_context_value("tool_output", tool_output.content())
}

#[cfg(test)]
mod test {
    use serde_json::json;

    use crate::agent::tools::search_file;

    use super::*;

    #[tokio::test]
    async fn test_prompt_rendering() {
        let tool = search_file();

        let tool_call = ToolCall::builder()
            .name("search_file")
            .id("1")
            .args(json!({ "query": "some_file"}).to_string())
            .build()
            .unwrap();
        let tool_output = ToolOutput::Text("Found it!".into());

        let available_tools = vec![Box::new(tool) as Box<dyn Tool>];
        let diff = indoc::indoc!(
            "
            diff --git a/some_file b/some_file
            index 0000000..1111111 100644
            --- a/some_file
            +++ b/some_file
            @@ -1,1 +1,1 @@
            -old
            +new
                "
        );

        let rendered_prompt = prompt(
            &available_tools[0],
            &tool_call,
            &tool_output,
            &available_tools,
        )
        .with_context_value("diff", diff);

        insta::assert_snapshot!(rendered_prompt.render().await.unwrap());
    }

    #[tokio::test]
    async fn test_prompt_render_where_tool_output_has_jinja() {
        // Create a mock tool
        let tool = search_file();

        // Create a ToolCall for this tool
        let tool_call = ToolCall::builder()
            .name("search_file")
            .id("2")
            .args(json!({ "query": "some_file_with_jinja"}).to_string())
            .build()
            .unwrap();

        // Tool output contains Jinja syntax
        let tool_output = ToolOutput::Text("Result with Jinja syntax: {{ some_variable }}".into());

        // Setup available tools
        let available_tools = vec![Box::new(tool) as Box<dyn Tool>];

        // Generate the prompt
        let rendered_prompt = prompt(
            &available_tools[0],
            &tool_call,
            &tool_output,
            &available_tools,
        )
        .with_context_value("diff", None::<String>);

        // Verify the output to ensure Jinja is escaped
        insta::assert_snapshot!(rendered_prompt.render().await.unwrap());
    }
}
