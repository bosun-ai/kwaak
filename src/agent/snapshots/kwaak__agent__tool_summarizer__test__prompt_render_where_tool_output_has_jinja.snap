---
source: src/agent/tool_summarizer.rs
expression: rendered_prompt.render().await.unwrap()
---
# Goal

A coding agent has made a tool call. It is your job to refine the output.

Reformat the following tool output such that it is effective for an AI agent to work with. Only include the reformatted output in your response.
Reformat but do not summarize, all information should be preserved and detailed.

##

Tool name: search_file
Tool description: Searches for a file inside the current project, leave the argument empty to list all files. Uses `find`.
Tool was called with arguments: {"query":"some_file_with_jinja"}



## Tool output

````shell
Result with Jinja syntax: {{ some_variable }}
````

## Format

- Only include the reformatted output in your response and nothing else.
- Include clear instructions on how to fix each issue using the tools that are
  available only.
- If you do not have a clear solution, state that you do not have a clear solution.
- If there is any mangling in the tool response, reformat it to be readable.


## Available tools

- **search_file**: Searches for a file inside the current project, leave the argument empty to list all files. Uses `find`.

## Requirements

- Only propose improvements that can be fixed by the tools and functions that are
  available in the conversation. For instance, running a command to fix linting can also be fixed by writing to that file without errors.
- If the tool output has repeating patterns, only include the pattern once and state
  that it happens multiple times.
- You cannot call tools yourself. You can only propose changes that can be made by
  the agent.
